File: server/src/ml/test_predict.py
#!/usr/bin/env python3
import os
import sys
import json
import subprocess

# Sample data for prediction
sample_data = {
    "Category": "Work",
    "Action": "Fix errors in code",
    "Priority": "High",
    "EstimatedTime": 60,
    "ActualTime": 90,
    "CompletionPercentage": 0.8,
    "TimeSpent": 75,
    "Status": "Over"
}

# Get the path to the predict.py script
script_dir = os.path.dirname(os.path.abspath(__file__))
predict_script = os.path.join(script_dir, 'predict.py')

print(f"Running predict.py at {predict_script}")

# Run the predict.py script as a subprocess
process = subprocess.Popen(
    [sys.executable, predict_script],
    stdin=subprocess.PIPE,
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE,
    text=True
)

# Send the sample data to the script
input_json = json.dumps(sample_data)
stdout, stderr = process.communicate(input_json)

# Print the output
print("\nStandard Output:")
print(stdout)

if stderr:
    print("\nStandard Error:")
    print(stderr)

print(f"\nExit Code: {process.returncode}")

# Try to parse the JSON output
try:
    result = json.loads(stdout)
    print("\nParsed Result:")
    print(f"Prediction: {result.get('prediction')}")
    print(f"Confidence: {result.get('confidence')}")
except json.JSONDecodeError as e:
    print(f"\nError parsing JSON: {e}")
except Exception as e:
    print(f"\nUnexpected error: {e}") 

File: server/src/ml/test_predict.js
// Test script for predict.py
import { spawn } from 'child_process';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Sample input data for prediction
const sampleData = {
  Category: 'Work',
  Action: 'Fix errors in code',
  Priority: 'High',
  EstimatedTime: 60,
  ActualTime: 90,
  CompletionPercentage: 0.8,
  TimeSpent: 75,
  Status: 'Over'
};

// Path to predict.py
const pythonScriptPath = path.join(__dirname, 'predict.py');
console.log(`Running Python script: ${pythonScriptPath}`);

// Run the Python script
const pythonProcess = spawn('python3', [pythonScriptPath]);

let stdoutBuffer = '';
let stderrBuffer = '';

// Handle Python script output
pythonProcess.stdout.on('data', (data) => {
  stdoutBuffer += data.toString();
  console.log('Python stdout:', data.toString());
});

// Handle Python script errors
pythonProcess.stderr.on('data', (data) => {
  stderrBuffer += data.toString();
  console.log('Python stderr:', data.toString());
});

// Handle Python script exit
pythonProcess.on('close', (code) => {
  console.log(`Python process exited with code ${code}`);
  
  if (code === 0) {
    // Try to extract the JSON from the stdout buffer
    try {
      // Look for a valid JSON object in the output
      const jsonMatch = stdoutBuffer.match(/(\{.*\})/);
      if (jsonMatch) {
        const jsonStr = jsonMatch[0];
        const result = JSON.parse(jsonStr);
        console.log('Parsed prediction result:', result);
        console.log(`Quality: ${result.prediction}, Confidence: ${(result.confidence * 100).toFixed(2)}%`);
      } else {
        console.error('No valid JSON found in the output');
      }
    } catch (e) {
      console.error('Failed to parse JSON:', e);
      console.error('Raw stdout output:', stdoutBuffer);
    }
  } else {
    console.error('Python script failed with exit code:', code);
    console.error('Error output:', stderrBuffer);
  }
});

// Send input data to the Python script
pythonProcess.stdin.write(JSON.stringify(sampleData));
pythonProcess.stdin.end(); 